design/ceph/ceph-orchestrator-osd-add-remove.md

```diff
@@ -0,0 +1,169 @@
+# Adding and removing Ceph osds via Ceph orchestrator interface
+
+## Goals
+
+1. Give Ceph's Rook Orchestrator Plugin the ability to add and remove OSDs to/from a Rook-Ceph
+   cluster running in Kubernetes.
+2. Users of the Ceph Orchestrator CLI and/or Ceph Dashboard should be able to
+   create OSDs on nodes with as much complexity as Ceph is able.
+3. Do not introduce unnecessary complexity into Rook or into Rook's CephCluster CRD.
+
+## Proposal
+
+### Removing OSDs
+Removing OSDs is the simplest place to start with this design. When removing an OSD or OSDs via the
+Rook Orchestrator Module, the Ceph mgr running in Kubernetes will run a Kubernetes Job that will run
+the required `ceph-volume` command necessary to remove the OSD(s) from a node as desired by the Ceph
+user. The Orchestrator Module will also perform whatever Ceph operations are necessary to remove the
+OSD as would be expected by the user. This will replace the manual, imperative steps currently
+required by Rook-Ceph users to remove OSDs from the storage cluster.
+
+#### OSD removal in the CephCluster resource
+The CephCluster CRD allows adding disks to nodes in several places including disks on all nodes in
+the Kubernetes cluster, to a specific node, or via disk pattern matching (in two different ways). To
+avoid the complexity of having to manage the logic for determining how to remove a node from the
+CephCluster resource, OSD removal will **not** alter the CephCluster CRD. While doing so would be
+convenient for users, there are a lot of opportunities in development to miss corner cases or
+introduce logic bugs. This may mean that Rook could re-add the disk after the disk is removed via
+the Ceph Orchestrator interface if the CephCluster CRD is configured to do so, and the user should
+be made aware of this possibility via documentation.
```

Comment:
Agreed on this approach to separate the steps of OSD removal from updating the CR from the orchestrator CLI perspective. However, I would expect dashboard features built around this to be smart enough to update the CephCluster CR and then remove the OSDs that are no longer relevant. Otherwise, dashboard users would be very confused if OSDs re-appeared.

Reply:
I can certainly see conflicts if `useAllDevices: true`  and then we remove `osd.$ID`. If we zap that OSD and orchestration are trigger Rook will add it again.
If we don't zap the disk, I don't see any risk.

Reply:
In my view the ceph cluster CRD must be always in sync with the real cluster. It seems something forced and completely against the concept of "declarative" behavior in k8s.
Maybe we should redesign the way that  OSD's are specified in the CRD to make it compatible with the [drive groups specification](https://github.com/ceph/ceph/blob/master/src/python-common/ceph/deployment/drive_group.py)

Reply:
> I can certainly see conflicts if `useAllDevices: true` and then we remove `osd.$ID`. If we zap that OSD and orchestration are trigger Rook will add it again.
> If we don't zap the disk, I don't see any risk.

I will still maintain with some more detail here that removal should not alter the CephCluster resource's `useAllNodes`, `deviceFilter`, `devicePathFilter`, or any `devices` specified at the cluster level. Said differently, removal should only focus on `devices` specified under individual `nodes`. Otherwise, this risks breaking the declarative desires a user has for multiple nodes in the cluster.

Reply:
I didn't include this in my latest update, but we can discuss whether it makes sense to add some sort of OSD blacklist to blacklist particular disks (by ID?) from a Rook-Ceph cluster. I'm inclined to treat this as a separate idea/feature for later.

Reply:
In my view the cluster CRD should reflect the current status of the cluster. If we have exceptions to this basic principle of the declarative way to work in k8s then we are going to have multiple collateral issues. 

For example: We have `"useAllNodes, useAllDevices, deviceFilter, and devicePathFilter" `attributes that makes really easy and simple to configure OSD's in a cluster. But the negative part is that you do not know reading the CRD what OSD's are included in the cluster, so you do not have a real vision of your cluster, and this can drive to undesired or unexpected situations
Ex: 
- User adds a new hard disk and this is used as OSD if useAllDevices=True),
- you can delete properly an OSD because Rook will create it again.
- we need "tricks" in order to have the right functionality (OSD blacklist proposal)

I think that we need to have in the cluster CRD each of the OSD's in the cluster, with the attributes that will allow us to provision the OSD using ceph-volume and start the Pod.
This makes easy adding/removing OSD's in the cluster and without collateral unexpected consequences.

In my opinion the attributes `"useAllNodes, useAllDevices, deviceFilter, and devicePathFilter"`  should be included in a special section of the cluster CRD for automatic OSD provision. 
A new bool attribute  "automatic_provision" should indicate if Rook should create/remove OSD's using this attributes or not. 
If automatic provision is enabled , we shouldn't allow to add/remove single OSD's , because this functionality depends of the selection of devices produced by the combination of `"useAllNodes, useAllDevices, deviceFilter, and devicePathFilter"` attributes.

If the user changes  "automatic_provision"  to False, then we should generate a complete "map" of the current OSD's in the cluster CRD. , and once we have this "map" we should allow to delete/create single OSD's.